src/car_backed_blockstore.rs

```diff
@@ -0,0 +1,333 @@
+//! │body length│frame body   │
+//! └───────────┼─────────────┤
+//!             │             │
+//! frame body ►│◄───────────►│
+//!     offset     =body length
+//! ```
+//!
+//! The first varint frame is a _header frame_, where the frame body is a [`CarHeader`] encoded using [`ipld_dagcbor`](serde_ipld_dagcbor).
+//!
+//! Subsequent varint frames are _block frames_, where the frame body is a concatenation of a [`Cid`] and the _block data_ addressed by that CID.
+//!
+//! ```text
+//! block frame ►│
+//! body offset  │
+//!              │  =body length
+//!              │◄────────────►│
+//!  ┌───────────┼───┬──────────┤
+//!  │body length│cid│block data│
+//!  └───────────┴───┼──────────┤
+//!                  │◄────────►│
+//!                  │  =block data length
+//!      block data  │
+//!          offset ►│
+//! ```
+//!
+//! ## Block ordering
+//! > _... a filecoin-deterministic car-file is currently implementation-defined as containing all DAG-forming blocks in first-seen order, as a result of a depth-first DAG traversal starting from a single root._
+//!
+//! - [CAR documentation](https://ipld.io/specs/transport/car/carv1/#determinism)
+//!
+//! # Future work
+//! - [`fadvise`](https://linux.die.net/man/2/posix_fadvise)-based APIs to pre-fetch parts of the file, to improve random access performance.
+//! - Use an inner [`Blockstore`] for writes.
+//! - Support compressed snapshots.
+//!   Note that [`zstd`](https://github.com/facebook/zstd/blob/e4aeaebc201ba49fec50b087aeb15343c63712e5/doc/zstd_compression_format.md#zstandard-frames) archives are also composed of frames.
+//!   Snapshots typically comprise of a single frame, but that would require decompressing all preceding data, precluding random access.
+//!   So compressed snapshot support would compression per-frame, or maybe per block of frames.
+//! - Support multiple files by concatenating them.
+//! - Use safe arithmetic for all operations - a malicious frame shouldn't cause a crash.
+//! - Theoretically, [`CarBackedBlockstore`] should be clonable (or even [`Sync`]) with very low overhead, so that multiple threads could perform operations concurrently.
+
+use ahash::AHashMap;
+use cid::Cid;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_car::CarHeader;
+use parking_lot::Mutex;
+use std::collections::hash_map::Entry::{Occupied, Vacant};
+use std::io::{
+    self,
+    ErrorKind::{InvalidData, Other, UnexpectedEof, Unsupported},
+    Read, Seek, SeekFrom,
+};
+use tracing::{debug, trace};
+
+/// If you seek to `offset` (from the start of the file), and read `length` bytes,
+/// you should get data that corresponds to a [`Cid`] (but NOT the [`Cid`] itself).
+#[derive(Debug)]
+struct BlockDataLocation {
+    offset: u64,
+    length: u32,
+}
+
+/// See [module documentation](mod@self) for more.
+pub struct CarBackedBlockstore<ReaderT> {
+    // https://github.com/ChainSafe/forest/issues/3096
+    inner: Mutex<CarBackedBlockstoreInner<ReaderT>>,
```

Comment:
@lemmih did you have to do something similar, or am I missing something?

Reply:
See also #2676

Reply:
Yeah, I had to do something similar.